service-loadbalancer/gce/controller.go

```diff
@@ -0,0 +1,295 @@
+	nodeController *framework.Controller
+	svcController  *framework.Controller
+	ingLister      StoreToIngressLister
+	nodeLister     cache.StoreToNodeLister
+	svcLister      cache.StoreToServiceLister
+	clusterManager *ClusterManager
+	recorder       record.EventRecorder
+	nodeQueue      *taskQueue
+	ingQueue       *taskQueue
+	tr             *gceTranslator
+	stopCh         chan struct{}
+	// stopLock is used to enforce only a single call to Stop is active.
+	// Needed because we allow stopping through an http endpoint and
+	// allowing concurrent stoppers leads to stack traces.
+	stopLock sync.Mutex
+	shutdown bool
+}
+
+// NewLoadBalancerController creates a controller for gce loadbalancers.
+func NewLoadBalancerController(kubeClient *client.Client, clusterManager *ClusterManager, resyncPeriod time.Duration) (*loadBalancerController, error) {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+
+	lbc := loadBalancerController{
+		client:         kubeClient,
+		clusterManager: clusterManager,
+		stopCh:         make(chan struct{}),
+		recorder: eventBroadcaster.NewRecorder(
+			api.EventSource{Component: "loadbalancer-controller"}),
+	}
+	lbc.nodeQueue = NewTaskQueue(lbc.syncNodes)
+	lbc.ingQueue = NewTaskQueue(lbc.sync)
+
+	// Ingress watch handlers
+	pathHandlers := framework.ResourceEventHandlerFuncs{
+		AddFunc:    lbc.ingQueue.enqueue,
+		DeleteFunc: lbc.ingQueue.enqueue,
+		UpdateFunc: func(old, cur interface{}) {
+			if !reflect.DeepEqual(old, cur) {
+				glog.Infof("Ingress %v changed, syncing",
+					cur.(*experimental.Ingress).Name)
+			}
+			lbc.ingQueue.enqueue(cur)
+		},
+	}
+	lbc.ingLister.Store, lbc.ingController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return lbc.client.Experimental().Ingress(api.NamespaceAll).List(
+					labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return lbc.client.Experimental().Ingress(api.NamespaceAll).Watch(
+					labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&experimental.Ingress{}, resyncPeriod, pathHandlers)
+
+	// Service watch handlers
+	svcHandlers := framework.ResourceEventHandlerFuncs{
+		AddFunc: lbc.enqueueIngressForService,
+		UpdateFunc: func(old, cur interface{}) {
+			if !reflect.DeepEqual(old, cur) {
+				lbc.enqueueIngressForService(cur)
+			}
+		},
+		// Ingress deletes matter, service deletes don't.
+	}
+
+	lbc.svcLister.Store, lbc.svcController = framework.NewInformer(
+		cache.NewListWatchFromClient(
+			lbc.client, "services", api.NamespaceAll, fields.Everything()),
+		&api.Service{}, resyncPeriod, svcHandlers)
+
+	nodeHandlers := framework.ResourceEventHandlerFuncs{
+		AddFunc:    lbc.nodeQueue.enqueue,
+		DeleteFunc: lbc.nodeQueue.enqueue,
+		// Nodes are updated every 10s and we don't care, so no update handler.
```

Comment:
We don't trust the event that's sent on the wire for this reason, the Informer will call the add handler if the thing it got isn't in local store (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/framework/controller.go#L246). If we get some add/delete/drop update pattern that snubs us the relist will eventually kick in, whenever.